### DISCUSSION TITLE
Send error message without context from command

### DISCUSSION THREAD

**Discussion by Drakan1** [Rapptz/discord.py#9990] [Q&A] [answered]
Currently, I have a few options to alert the user to errors in the code.
    The first is the logger module, works fine no issues. But as I don't check the logs without a heads up, I also appreciate a discord message.
    The second is using the context given from a command. This is fine for higher level methods/classes but for things that do mostly data manipulation they currently don't get passed or deal with any discord objects to access the context variable.
    The third is using a raise commandError. I have overwritten on_command_error such that when a commandError is sent, it does a logger.warning and sends a discord message with the error using the context. This is great but using a raise, breaks the flow of execution which sometimes you don't want.
    (E.g. have a method that executes last in a command, update_data, with an arg which tells what data to send to the db.)
    
 So the question is this, how can I send a message to the user with an error for those low level functions without breaking the execution flow?
    For those cases I was happy exiting out the lower function but still wanted higher level logic, I still used the raise but wrapped the function with a `with suppress (commandError)` but then you lose the message.
    The hope was I could access the event loop somehow to see the parent command's context that led to that function. `bot.get_context()` requires an object argument to get the context from so that doesn't work for me either.
    So it only leaves me with either temporarily storing the context variable somewhere (but doesn't really work for multi user/guild situations ). In the past I have got around this by just having a guild specific error channel but less practical as opening the bot to more users in more channels. Otherwise I can crudely pass the context into every function / method when it doesn't belong.
    Is there a better way?

**Comment by Drakan1** [comment, marked as answer]
I solved it by using the frame stack:

```py
def _get_frame_context(self, logger=None):
    """Inspect the stack to find the context object from invoke"""
    frame_stack = inspect_stack()
    for frame_info in frame_stack:
        if frame_info.function.lower() == "invoke":
            context = frame_info.frame.f_locals["ctx"]
            break
    else:
        logger = self.logger if not logger else logger
        raise CommandError("Could not find context object", logger)
    del frame_stack
    return context

async def send_error(self, message: str, logger: logging.Logger, context=None):
    """
    :param message: The error message
    :param logger: The logger from where the error occured

    Send a message to the invocation channel, without disrupting execution
    if can find context
    """
    if context is None:
        try:
            context = await self._get_frame_context()
        except CommandError:
            raise CommandError(f"Could not find context object: {message}", logger)

    await context.send(message)
    logger.warning(message)
    ```

**Reply by Soheab** [reply]
I don’t understand your original problem, but you could have used a [webhook](https://support.discord.com/hc/en-us/articles/228383668-Intro-to-Webhooks) to send to a channel without any state:
```py
import discord

web = discord.Webhook.from_url("webhook_url", session=<aiohttp.ClientSession>)
# or Webhook.partial(token=..., id=..., session=...)
await web.send(...)
```
https://discordpy.readthedocs.io/en/stable/api.html#discord.Webhook

**Reply by Drakan1** [reply]
The problem was I wanted to send a message to the channel where the user sent the command, not just an arbitrary channel
E.g.
1.[User] !ping
2.[Bot] We encountered an error

Obviously, in a more complex situation where the `ctx` is not freely available
